### PR TITLE
ZK-4745: Datebox supports to display with custom calendar system (e.g…

### PR DESCRIPTION
**Pull request by plijyfes**
…. Minguo) year

### CODE REVIEW

File: zk/src/org/zkoss/zk/ui/http/Wpds.java [hunk 1]
```diff
@@ -233,13 +233,13 @@ private static final String getDateJavaScript(Locale locale, int firstDayOfWeek)
 		df = new SimpleDateFormat("G", locale);
 		final boolean helang = locale.getLanguage().equals("iw");
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/ui/http/Wpds.java:234]: remove

File: zk/src/org/zkoss/zk/ui/http/Wpds.java [hunk 2]
```diff
@@ -233,13 +233,13 @@ private static final String getDateJavaScript(Locale locale, int firstDayOfWeek)
...
-			sb.append("zk.ERA='").append(df.format(new java.util.Date())).append("';\n");
+			sb.append("zk.ERA='").append("';\n");
 		else
-			sb.append("zk.ERA=\"").append(df.format(new java.util.Date())).append("\";\n");
+			sb.append("zk.ERA=\"").append("\";\n");
 
-		Calendar ec = Calendar.getInstance(Locale.ENGLISH);
-		Calendar lc = Calendar.getInstance(locale);
-		sb.append("zk.YDELTA=").append(lc.get(Calendar.YEAR) - ec.get(Calendar.YEAR)).append(";\n");
+		sb.append("zk.YDELTA=").append(0).append(";\n");
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/ui/http/Wpds.java:240]: merge append

File: zul/src/archive/web/js/zul/db/Calendar.js [hunk 3]
```diff
@@ -321,6 +331,48 @@ zul.db.Renderer = {
...
+		return result || localeDateString.split(' ')[0];
+	},
+	_getYDelta(date, langTag) {
+		var loacleCalendarYear = date.toLocaleDateString(langTag, {year: 'numeric'}).replace(/\D/g, ''),
+			isoCalendarYear = date.toLocaleDateString('en-US', {year: 'numeric'}).replace(/\D/g, '');
+		return loacleCalendarYear - isoCalendarYear;
+	},
+	_getDisplayYear(date, localizedSymbols) {
+		var year = date.getFullYear();
+		if (!zk.feature.ee) return year; // ZK-4745: displaying specific calendar system is for ZK EE only
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/Calendar.js:370]: block at server side

File: zul/src/archive/web/js/zul/db/Calendar.js [hunk 4]
```diff
@@ -321,6 +331,48 @@ zul.db.Renderer = {
...
+		var langTag = localizedSymbols.LAN_TAG,
+			ydelta = this._getYDelta(date, langTag),
+			era = (langTag.indexOf('-ca-') != -1) ? this._getEraName(date, localizedSymbols) : '';
+		year = (year + ydelta > 0) ? year + ydelta : 1 - year - ydelta;
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/Calendar.js:374]: extract year + ydelta

File: zul/src/archive/web/js/zul/db/Calendar.js [hunk 5]
```diff
@@ -321,6 +331,48 @@ zul.db.Renderer = {
...
+				if (eraYearGap !== undefined && eraYearMatched > eraYearGap) {
+					eraYearMatched = eraYearGap;
+					result = key;
+				}
+			}
+		}
+		return result || localeDateString.split(' ')[0];
+	},
+	_getYDelta(date, langTag) {
+		var loacleCalendarYear = date.toLocaleDateString(langTag, {year: 'numeric'}).replace(/\D/g, ''),
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/Calendar.js:364]: typo

File: zul/src/archive/web/js/zul/db/Calendar.js [hunk 6]
```diff
@@ -321,6 +331,48 @@ zul.db.Renderer = {
...
+				if (localeDateString.indexOf(key) != -1)
+					return key; // match with localeDateString
+
+				// client era name not match localizedSymbols, need to calculate correct era name with year and eras
+				var eraFirstYear = eras[key].firstYear,
+					isPositive = eras[key].direction > 0,
+					eraYearGap;
+				if ((isPositive && year >= eraFirstYear) || (!isPositive && year <= eraFirstYear))
+					eraYearGap = Math.abs(year - eraFirstYear);
+				if (eraYearGap !== undefined && eraYearMatched > eraYearGap) {
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/Calendar.js:355]: eraYearMatched  === undefined || eraYearMatched > eraYearGap
and remove 9999

Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/Calendar.js:355]: no need eraYearGap !== undefined

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 7]
```diff
@@ -202,12 +239,30 @@ zk.fmt.Date = {
...
+			for (var i = 0; i < len; i++) {
+				var key = erakeys[i];
+				if (txt.indexOf(key) != -1) {
+					era = eras[key];
+					eraName = key;
+					break;
+				}
+			}
+		}
+		var ydelta = localizedSymbols.YDELTA || (era ? 0 - era.firstYear + era.direction : _getYDelta(refval, localizedSymbols.LAN_TAG));
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:265]: 0 - era.firstYear + era.direction * 1

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 8]
```diff
@@ -244,18 +300,20 @@ zk.fmt.Date = {
 
 					if (!isNaN(nv = _parseInt(token))) {
 						var newY = Math.min(nv, 200000); // Bug B50-3288904: js year limit
-						if (newY < 100) {
-							var fullYear = y + localizedSymbols.YDELTA;
-							if (newY === fullYear % 100) { // assume yy is not modified
-								y = fullYear;
-								break;
-							}
+						if (!hasG && newY < 100) { // ZK-4745: no need handle yy with G(era)
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:303]: try JAVA parse Gyy

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 9]
```diff
@@ -459,8 +528,14 @@ zk.fmt.Date = {
 
 				switch (cc) {
 				case 'y':
-					if (len <= 3) txt += _digitFixed(val.getFullYear() % 100, 2);
-					else txt += _digitFixed(val.getFullYear(), len);
+					var y,
+						localeDateString = val._moment.toDate().toLocaleDateString(localizedSymbols.LAN_TAG, {era: 'short', year: 'numeric'});
+					if (!/\d/.test(localeDateString) && localeDateString.length) // Could be the first year of era
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:533]: (e.g. 令和元年)

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 10]
```diff
@@ -459,8 +528,14 @@ zk.fmt.Date = {
...
+						y = 1;
+					else
+						y = parseInt(localeDateString.replace(/\D/g, ''));
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:536]: if parseInt get NaN then y = 1

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 11]
```diff
@@ -571,45 +647,38 @@ zk.fmt.Calendar = zk.$extends(zk.Object, {
...
 			if (val.getMonth() == 1 && val.getDate() == 29) {
 				d = new LeapDay(val); // a proxy of Date
 				d.setOffset(this._offset);
-			} else {
-				d = Dates.newInstance(val);
-				d.setFullYear(d.getFullYear() + this._offset);
 			}
 		}
-		return zk.fmt.Date.formatDate(d || val, fmt, localizedSymbols);
+		return zk.fmt.Date.formatDate(val, fmt, localizedSymbols);
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:658]: d || val

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 12]
```diff
@@ -540,6 +566,16 @@ zk.fmt.Date = {
...
+	},
+	getYDelta: function (date, langTag) { // for zkmax override
+		return 0;
+	},
+	getEraName: function (date, localizedSymbols) { // for zkmax override
+		var langTag = localizedSymbols.LAN_TAG.split('-u-')[0],
+			localeDateString = date.toLocaleDateString(langTag, {era: 'short', day: 'numeric'});
+		return localeDateString.split(' ')[0];
+	},
+	_parseEras: function (txt, eras) { // for zkmax override
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:578]: return null;

File: zul/src/archive/web/js/zul/db/Calendar.js [hunk 13]
```diff
@@ -321,6 +331,9 @@ zul.db.Renderer = {
 			val = new zk.fmt.Calendar().formatDate(zUtl.today(parent, tz), wgt.getFormat(), localizedSymbols);
 		}
 		out.push(val);
+	},
+	_getDisplayYear(date, localizedSymbols) { // for zkmax override
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/Calendar.js:335]: : function

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 14]
```diff
@@ -405,9 +430,9 @@ zk.fmt.Date = {
 		if (hasHour1 && isAM === false)
 			hr += 12;
 		var dt;
-		if (m == 1 && d == 29 && localizedSymbols.YDELTA) {
-			dt = new LeapDay(y - localizedSymbols.YDELTA, m, d, hr, min, sec, msec, tz);
-			dt.setOffset(localizedSymbols.YDELTA);
+		if (m == 1 && d == 29 && ydelta) {
+			dt = new LeapDay(y, m, d, hr, min, sec, msec, tz);
+			dt.setOffset(ydelta);
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:435]: check offset is needed

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 15]
```diff
@@ -459,8 +485,11 @@ zk.fmt.Date = {
 
 				switch (cc) {
 				case 'y':
-					if (len <= 3) txt += _digitFixed(val.getFullYear() % 100, 2);
-					else txt += _digitFixed(val.getFullYear(), len);
+				case 'u':
+					var ydelta = cc == 'y' ? localizedSymbols.YDELTA || this.getYDelta(val._moment.toDate(), localizedSymbols.LAN_TAG) : 0,
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:489]: cc == 'u' ydelta = 0;

File: zul/src/archive/web/js/zul/db/datefmt.js [hunk 16]
```diff
@@ -540,6 +566,16 @@ zk.fmt.Date = {
 			}
 		}
 		return txt;
+	},
+	getYDelta: function (date, langTag) { // for zkmax override
+		return 0;
+	},
+	getEraName: function (date, localizedSymbols) { // for zkmax override
+		var langTag = localizedSymbols.LAN_TAG.split('-u-')[0],
```
Comment (plijyfes) [inline zul/src/archive/web/js/zul/db/datefmt.js:574]: remove after second dash

**Review by scribetw — APPROVED** [review]

File: zul/src/archive/web/js/zul/db/Calendar.js [hunk 17]
```diff
@@ -140,42 +140,54 @@ zul.db.Renderer = {
 			val = wgt.getTime(),
 			m = val.getMonth(),
 			y = val.getFullYear(),
-			ydelta = new zk.fmt.Calendar(val, wgt._localizedSymbols).getYear() - y,
+			// to avoid moment using the last day(according to local timezone) of the previous year in ie.
+			date = zk.ie ? new Date(y, m) : val._moment.toDate(),
+			localeDateTimeFormat = zk.ie < 11 ? null : new Intl.DateTimeFormat(localizedSymbols.LAN_TAG, {year: 'numeric'}),
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/db/Calendar.js:145]: should note ie only supports 11

**Review by scribetw — APPROVED** [review]

**Review by DevChu — APPROVED** [review]

**Pull request merged**